pail/s3_bucket.go

```diff
@@ -0,0 +1,551 @@
+	return &largeWriteCloser{
+		maxSize: 5000000,
+		name:    s.name,
+		svc:     s.svc,
+		ctx:     ctx,
+		key:     s.normalizeKey(key),
+	}, nil
+}
+
+func (s *s3Bucket) Reader(ctx context.Context, key string) (io.ReadCloser, error) {
+	input := &s3.GetObjectInput{
+		Bucket: aws.String(s.name),
+		Key:    aws.String(s.normalizeKey(key)),
+	}
+
+	result, err := s.svc.GetObjectWithContext(ctx, input)
+	if err != nil {
+		return nil, err
+	}
+	return result.Body, nil
+}
+
+func putHelper(b Bucket, ctx context.Context, key string, r io.Reader) error {
+	f, err := b.Writer(ctx, key)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	_, err = io.Copy(f, r)
+	if err != nil {
+		_ = f.Close()
+		return errors.Wrap(err, "problem copying data to file")
+	}
+	return errors.WithStack(f.Close())
+}
+
+func (s *s3BucketSmall) Put(ctx context.Context, key string, r io.Reader) error {
+	return putHelper(s, ctx, key, r)
+}
+
+func (s *s3BucketLarge) Put(ctx context.Context, key string, r io.Reader) error {
+	return putHelper(s, ctx, key, r)
+}
+
+func (s *s3Bucket) Get(ctx context.Context, key string) (io.ReadCloser, error) {
+	return s.Reader(ctx, key)
+}
+
+func uploadHelper(b Bucket, ctx context.Context, key, path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return errors.Wrapf(err, "problem opening file %s", path)
+	}
+	defer f.Close()
+
+	return errors.WithStack(b.Put(ctx, key, f))
+}
+
+func (s *s3BucketSmall) Upload(ctx context.Context, key, path string) error {
+	return uploadHelper(s, ctx, key, path)
+}
+
+func (s *s3BucketLarge) Upload(ctx context.Context, key, path string) error {
+	return uploadHelper(s, ctx, key, path)
+}
+
+func (s *s3Bucket) Download(ctx context.Context, key, path string) error {
+	reader, err := s.Reader(ctx, key)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err = os.MkdirAll(filepath.Dir(path), 0700); err != nil {
+		return errors.Wrapf(err, "problem creating enclosing directory for '%s'", path)
+	}
+
+	f, err := os.Create(path)
+	if err != nil {
+		return errors.Wrapf(err, "problem creating file '%s'", path)
+	}
+	_, err = io.Copy(f, reader)
+	if err != nil {
+		_ = f.Close()
+		return errors.Wrap(err, "problem copying data")
+	}
+
+	return errors.WithStack(f.Close())
+}
+
+func (s *s3Bucket) pushHelper(b Bucket, ctx context.Context, local, remote string) error {
+	files, err := walkLocalTree(ctx, local)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	for _, fn := range files {
+		target := filepath.Join(remote, fn)
+		file := filepath.Join(local, fn)
+		localmd5, err := md5sum(file)
+		if err != nil {
+			return errors.Wrapf(err, "problem checksumming '%s'", file)
+		}
+		input := &s3.HeadObjectInput{
+			Bucket:  aws.String(s.name),
+			Key:     aws.String(target),
+			IfMatch: aws.String(localmd5),
+		}
+		_, err = s.svc.HeadObjectWithContext(ctx, input)
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == "PreconditionFailed" || aerr.Code() == "NotFound" {
```

Comment:
are these strings constants?

Reply:
Unfortunately, there are only two error code constants in the s3 go sdk, the rest are generic strings:

> In this example the S3 GetObject API operation is used to request the contents of an object in S3. The example handles the NoSuchBucket and NoSuchKey error codes, printing custom messages to stderr. If any other error is received, a generic message is printed.

https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/handling-errors.html